File: CS/06_classes_objects/ClassesObjects.cs
using System;

namespace BasicCS.Chapter06
{
    internal class ClassesObjects
    {
        static void Main()
        {
            // 객체 생성 (Object instantiation)
            var p1 = new Person("Alice", 30);
            p1.Introduce();

            // 객체 이니셜라이저 (Object initializer)
            var p2 = new Person { Name = "Bob", Age = 25 };
            p2.Introduce();

            // record (C# 9.0+)
            var r1 = new Product("Laptop", 1200m);
            var r2 = r1 with { Name = "Tablet" };
            Console.WriteLine($"Record: {r1}, Modified: {r2}");

            // static 멤버 (Static members)
            Console.WriteLine($"Total persons: {Person.Count}");

            // readonly struct 예제
            var pt = new Point(3, 4);
            Console.WriteLine($"Point distance: {pt.DistanceFromOrigin()}");

            // extension method
            Console.WriteLine($"\"hello\" reversed: {"hello".Reverse()}");
        }
    }

    // 기본 클래스 (Basic class)
    class Person
    {
        // 자동 구현 속성 (Auto-implemented properties)
        public string Name { get; set; }
        public int Age { get; set; }

        // 읽기 전용 속성 (Read-only property)
        public bool IsAdult => Age >= 18;

        // 정적 필드 (Static field)
        public static int Count { get; private set; } = 0;

        // 생성자 (Constructor)
        public Person() { Name = ""; Age = 0; Count++; }

        public Person(string name, int age) : this()
        {
            Name = name;
            Age = age;
        }

        // 메서드 (Method)
        public void Introduce()
        {
            Console.WriteLine($"Hi, I'm {Name}, {Age} years old. Adult: {IsAdult}");
        }

        // 소멸자 (Finalizer)
        ~Person()
        {
            Count--;
        }
    }

    // record (C# 9.0+)
    record Product(string Name, decimal Price);

    // readonly struct
    readonly struct Point
    {
        public int X { get; }
        public int Y { get; }

        public Point(int x, int y)
        {
            X = x;
            Y = y;
        }

        public double DistanceFromOrigin() => Math.Sqrt(X * X + Y * Y);
    }

    // 확장 메서드 (Extension methods)
    static class StringExtensions
    {
        public static string Reverse(this string s)
        {
            char[] chars = s.ToCharArray();
            Array.Reverse(chars);
            return new string(chars);
        }
    }
}

File: VB/07-classes-objects/ClassesObjects.vb
Imports System

Module Program
    Sub Main()
        ' 객체 생성
        Dim p1 As New Person("Alice", 25)
        p1.Introduce()

        ' 속성 사용
        p1.Name = "Bob"
        p1.Age = 30
        Console.WriteLine($"이름: {p1.Name}, 나이: {p1.Age}")

        ' Shared 멤버
        Console.WriteLine($"총 인원: {Person.Count}")

        ' 자동 구현 속성
        Dim c1 As New Car("Tesla")
        c1.Speed = 100
        Console.WriteLine($"{c1.Model} 속도: {c1.Speed} km/h")

        ' 읽기 전용 속성
        Console.WriteLine($"생성 시간: {p1.CreatedAt}")
    End Sub
End Module

' 클래스 정의
Public Class Person
    ' Shared(정적) 필드
    Public Shared Count As Integer = 0

    ' 속성 (자동 구현)
    Public Property Name As String

    ' 속성 (수동 구현)
    Private _age As Integer
    Public Property Age As Integer
        Get
            Return _age
        End Get
        Set(value As Integer)
            If value >= 0 Then _age = value
        End Set
    End Property

    ' 읽기 전용 속성
    Public ReadOnly Property CreatedAt As Date

    ' 생성자
    Public Sub New(name As String, age As Integer)
        Me.Name = name
        Me.Age = age
        Me.CreatedAt = Date.Now
        Count += 1
    End Sub

    ' 메서드
    Public Sub Introduce()
        Console.WriteLine($"안녕하세요, 저는 {Name}이고 {Age}세입니다.")
    End Sub
End Class

' 자동 구현 속성 예제
Public Class Car
    Public Property Model As String
    Public Property Speed As Integer

    Public Sub New(model As String)
        Me.Model = model
    End Sub
End Class
